# 💡 **예외**

자바에서 문법에 맞지 않게 작성한 코드는 사전에 컴파일러에 의해 오류로 걸러지지만,

예외는 컴파일은 잘 되지만 실행 중에 발생하게 된다.

<br>

> **예외처리란?**
>
> 프로그램 실행 시 발생할 수 있는 예기치 못한 상황에 대한 코드를 작성하는 것
>
> 목적은 프로그램의 비정상적인 종료를 막고, 정상적인 실행을 유지하는 것

<br>

- **try-catch문**

  try 블록 내에서 예외가 발생한 경우,

1. 예외가 발생한 이후 문장은 실행되지 않는다
2. 예외와 일치하는 catch 블록이 있다면 문장을 수행한다
3. finally 블록은 항상 실행된다(생략 가능)

```java
public void arrayOutOfBound() {
	int[] intArray = new int[5];
	try {
		System.out.println(intArray[5]);
		System.out.println("This code should run."); // 실행되지 않음, try블럭에서 예외가 발생할 경우, 예외 발생 위치 이후의 문장들은 수행되지 않는다.
	} catch (Exception e) {
		System.err.println(intArray.length);
	} finally {
		System.out.println("finally is called");
	}

	System.out.println("This code must run.");
	}
```

```java
 > 5
 > finally is called
 > This code must run.
```

<br>

- **다중 catch블록**

  catch 블록은 순서를 따진다.
  Exception 클래스는 부모 클래스 이므로, 가장 아래에 놓여야 한다. 부모 클래스가 상위에서 정의되면 모든 예외를 처리하게 되므로 다음 catch 블록 부터는 비교 수행 할 수 없기 때문이다.

<br>

- **java.lang.Throwable 클래스**

  Error, Exception의 부모 클래스로 모든 예외 클래스들이 가지고 있는 공통된 메소드를 정의

  - getMessage() : 예외 메세지를 문자열로 반환

  - toString() : 간단한 설명을 반환(예외 클래스 이름 제공)

  - printStackTrace() : 예외가 발생하게 된 메소드들의 호출 관계 출력

<br>

- **예외의 책임 전가 throws**

  throws로 선언되어 있는 메소드를 **호출한 메소드**에서 try-catch 구문으로 처리함

  예외 클래스는 여러개일 수도 있다
  
 <br>
 
- **checked와 unchecked**

  부모 중 RuntimeException이 있을 경우엔 unchecked, 없을 경우엔 checked
  
  checked 예외의 경우엔 반드시 예외 처리를 해주어야 함




